File: c++leet/num_smaller_numbers/num_smaller.cpp
#include <unordered_map>
#include <iostream>
#include <stdio.h>
#include <string>
#include <vector>
#include <set>
using namespace std;

class Solution{
  public:
    vector<int> smallerNumbersThanCurrent(vector<int>& nums){
      vector<int> nums_sorted (nums.begin(), nums.end());
      vector<int> nums_smaller;
      unordered_map<int, int> index_map; 
      unordered_map<int, int>::iterator found;  
      int tmp;
      int counter=0;

      for (int i=0; i < nums_sorted.size()-1; i++){
        for (int j=(i+1); j < nums_sorted.size(); j++){
          if (nums_sorted[i] > nums_sorted[j]){
            tmp = nums_sorted[i];
            nums_sorted[i] = nums_sorted[j];
            nums_sorted[j] = tmp; 
          }
        }
      }
      for (auto n: nums){
        found = index_map.find(n);
        if (found == index_map.end()){
          while (nums_sorted[counter] != n){
            counter+=1;
          }
          index_map.insert({n, counter});
          counter = 0;
          found = index_map.find(n);
        }
        nums_smaller.push_back(found->second);
      } 
      return nums_smaller;
    }
};



int main(){
  vector<int> nums = {16,4,2,47,2,8,1,1,5,3};
  Solution sol;
  sol.smallerNumbersThanCurrent(nums);
  return 0;
}

File: How_Many_Numbers_Are_Smaller_Than_The_Current_Number/js.d/num_smaller.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function(nums) {
    var nums_sorted = [...nums];
    nums_sorted.sort((a, b) => a - b);
    var nums_smaller = [];
    var index_map = {};
    for (i=0; i < nums.length; i++){
      if (! index_map.hasOwnProperty(nums[i])){
        index_map[nums[i]] = nums_sorted.indexOf(nums[i]);
      }
      nums_smaller.push(index_map[nums[i]])
    }
    return nums_smaller;
};

console.log(smallerNumbersThanCurrent([5,0,10,0,10,6])) // => [2,0,4,0,4,3]
